source/unified-test-format/unified-test-format.rst

```diff
@@ -462,6 +477,78 @@ The structure of this object is as follows:
 
     Test files SHOULD NOT use this option unless one or more command monitoring
     events are specified in `observeEvents <entity_client_observeEvents_>`_.
+    
+  - ``storeEventsAsEntities``: Optional map of event names to entity names.
+    If provided the test runner MUST:
+    
+      - For each entity name, create the respective entity with a type of
+        "event list". If multiple event names map to the same entity name,
+        exactly one entity MUST be created. If the entity already exists
+        (such as from a previous ``storeEventsAsEntities`` declaration from
+        another client), the test runner MUST raise an error.
+      - Set up an event subscriber for each event named. The event subscriber
+        MUST serialize the events it receives into a document and append
+        the document to the array stored in the specified entity.
+        The fields in the document MUST be as follows:
+        
+        - For ``CommandStartedEvent``:
+       
+          - ``name``: the name of the event, i.e. ``CommandStartedEvent``.
+          - ``commandName``: the name of the command, e.g. ``insert``.
+          - ``startTime``: the (floating-point) number of seconds since
+            the Unix epoch when the command began executing.
```

Comment:
I added language explicitly requiring the test runner to correlate the started/succeeded and started/failed events.

Reply:
As mentioned in https://github.com/mongodb/specifications/pull/903/files#r567703784, if we're not going to confine this logic to a monolithic special operation, I think it's preferable to offload the logic for correlating CommandStarted and CommandSucceeded/Failed events into whatever consumes the entity map (i.e. workload executor).

Could this section be simplified to just refer to the event objects as defined in their respective specs?

 * Connection monitoring event objects: https://github.com/mongodb/specifications/blob/master/source/command-monitoring/command-monitoring.rst#api
 * CMAP event objects: https://github.com/mongodb/specifications/blob/master/source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst#events

On top of the specified fields, you can require the test runner to tack on `name` and `time` (or `observedAt`) fields to each record. We can also allow test runners to omit fields if they know they won't be needed by whatever consumes the entity. For instance, `command` and `reply` are huge and aren't relevant to the current design of the workload executor. However, `requestId` would be necessary for the workload executor to correlate command events.

If you're adamant about enumerating the fields and not merely referring to the other specs, that's fine.

Reply:
My goal primarily is to make whatever events are collected useful. Without correlating started and succeeded/failed events, I find either one individually to not particularly useful. Currently this correlation is primarily happening by hand by the person looking at event logs (as I understand it), but I think this manual correlation is less feasible for DAT due to the volume of events that could be produced.

Therefore, I think it would be valuable to have the events be pre-correlated by *something*.

As to what should perform the correlation, I think it should really be the driver when it publishes the events in the first place. But, since we are not changing the driver in the present project, the closest point would be the test runner (which is still in the driver's code base and hence the correlation code perhaps has a chance of being moved to the driver proper  down the road).

I am specifically having a hard time understanding why you propose the test runner performs filtering (which requires knowledge of events and a processing step) but not correlation which seems to be quite similar. Can you elaborate on why you don't want the test runner to correlate?

Reply:
Per our discussion, I removed command monitoring event storage, leaving only CMAP events, and changed the fields to be exactly the specified fields + name + observedAt.